pages/docs/v3/core-apis/data-types.md

```diff
@@ -0,0 +1,116 @@
+
+### Null Values
+
+Objective-C does not support storing null values in collections such as arrays, dictionaries, or sets. Instead it uses a special placeholder object, [`NSNull`](https://developer.apple.com/documentation/foundation/nsnull?language=objc), to represent a null value. In contrast, Swift uses [Optionals](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html) to describe a value that might be null. Swift can manipulate `NSNull` values but Objective-C cannot handle Optionals (although, in some contexts, the runtime will automatically map optionals into the underlying value or `NSNull`). These `NSNull` objects can appear regardless of which language you are using.
+
+As an example, consider the following object being passed to a Capacitor plugin call:
+
+```typescript
+{ 'foo': null, 'bar': [1, 2, null, 4]}
+```
+
+#### Dictionaries
+
+`CAPPluginCall` stores this data as its `options` dictionary property but has a variety of convenience accessors that operate on it. The convenience accessors will try to cast the value to the expected type(s) so `NSNull` values will get filtered out.
+
+```swift
+if let value = call.getString("foo") {
+    // false, `value` is nil
+}
+```
+
+However, accessing the storage property directly can return an `NSNull` object so its type will need to be checked before it is used.
+
+```swift
+let value = call.options["foo"]
+if let value = value {
+    // true, 'value' is a valid object
+}
+if let value = value as? Int {
+    // false, the NSNull object cannot be cast to Int
+}
+if let value = value as? NSNull {
+    // true, 'value' is a NSNull
+}
+```
+
+In general, it is not recommended to rely on the presence of a key to convey meaning. Only the corresponding value should be evaluated after its expected type is confirmed. Since dictionaries typically require the typing of each value as it is extracted, the inclusion of `NSNull` can mostly be ignored.
+
+#### Arrays
+
+Since accessing arrays typically requiring typing the whole collection, it is more important to consider if they are heterogeneous or not.
+
+```swift
+if let values = call.getArray("bar") {
+    // true, the array is all valid objects
+}
+if let values = call.getArray("bar", Int) {
+    // false, the array is a mix of Int and NSNull objects and can't be cast to [Int]
+}
+if let values = call.getArray("bar", Int?) {
+    // false, the array is a mix of Int and NSNull objects and can't be cast to [Int?]
+}
+```
+
+To help with this behavior, Capacitor includes a convenience extension that can map a heterogeneous array with `NSNull` values into an array of optionals. It works on the `JSValue` protocol, which represents all of the valid types that can be bridged between environments, but can be cast to a specific subtype.
+
+```swift
+if let values = call.getArray("bar").capacitor.replacingNullValues() as? [Int?] {
```

Comment:
Would it make sense if our `getArray` helper did this by default? Or maybe it can't because of Obj-C?

Reply:
`getArray` is Swift only so Obj-C is not a concern. The issue is that if we did that by default, then all arrays would get typed as optional regardless of whether or not they contain nulls. I'm trying to avoid optionals where possible to reduce the temptation of force unwrapping (which is why the accessors with a default value no longer return optionals). So it seemed better to put the burden on the less common use-case rather than make everyone cast/test the result.